File: examples/v1_4_0_features_demo.py
# Depyler v1.4.0 - Async/Await Support Demo

# 1. Basic async function
async def hello_async() -> str:
    return "Hello from async!"

# 2. Async function with await
async def fetch_user(user_id: int) -> dict[str, str]:
    # Simulate async database fetch
    await async_sleep(0.1)
    return {"id": str(user_id), "name": f"User{user_id}"}

async def async_sleep(seconds: float) -> None:
    # Placeholder for actual async sleep
    pass

# 3. Async methods in classes
class AsyncService:
    def __init__(self, name: str):
        self.name = name
        self.requests_count = 0
    
    async def handle_request(self, request_id: int) -> str:
        self.requests_count += 1
        result = await self._process(request_id)
        return f"{self.name} processed: {result}"
    
    async def _process(self, request_id: int) -> str:
        await async_sleep(0.05)
        return f"Request#{request_id}"

# 4. Async with multiple awaits
async def process_batch(items: list[int]) -> list[str]:
    results = []
    for item in items:
        user = await fetch_user(item)
        result = f"Processed {user['name']}"
        results.append(result)
    return results

# 5. Async main function
async def main() -> None:
    # Test basic async
    greeting = await hello_async()
    print(greeting)
    
    # Test async class methods
    service = AsyncService("API-Service")
    response = await service.handle_request(123)
    print(response)
    
    # Test batch processing
    items = [1, 2, 3, 4, 5]
    batch_results = await process_batch(items)
    for result in batch_results:
        print(result)

File: examples/v1_4_0_features_demo.rs
#![allow(unused_imports)]
#![allow(unused_mut)]
#![allow(unused_variables)]
#![allow(unreachable_patterns)]
#![allow(unused_assignments)]
#![allow(dead_code)]
use std::collections::HashMap;
    #[derive(Debug, Clone)] pub struct AsyncService {
    pub name: String, pub requests_count: i32
}
impl AsyncService {
    pub fn new(name: impl Into<String>) -> Self {
    Self {
    name: name.into(), requests_count: 0
}
} pub async fn handle_request(&mut self, request_id: i32) -> String {
    self.requests_count = self.requests_count.clone() + 1;
    let result = self._process(request_id).await;
    return format !("{} processed: {}", self.name.clone(), result);
   
}
pub async fn _process(&self, request_id: i32) -> String {
    async_sleep(0.05).await;
    return format !("Request#{}", request_id);
   
}
} #[doc = " Depyler: verified panic-free"] #[doc = " Depyler: proven to terminate"] pub async fn hello_async() -> String {
    "Hello from async!".to_string().to_string()
}
#[doc = " Depyler: verified panic-free"] #[doc = " Depyler: proven to terminate"] pub async fn fetch_user(user_id: i32) -> HashMap<String, String>{
    async_sleep(0.1).await;
    {
    let mut map: HashMap<String, String>= HashMap::new();
    map.insert("id".to_string() ,(user_id).to_string());
    map.insert("name".to_string(), format !("User{}", user_id));
    map
}
} #[doc = " Depyler: verified panic-free"] #[doc = " Depyler: proven to terminate"] pub async fn async_sleep(_seconds: f64) {
   
}
#[doc = " Depyler: verified panic-free"] pub async fn process_batch(items: & Vec<i32>) -> Vec<String>{
    let mut results: Vec<String>= vec ! [];
    for item in items.iter().cloned() {
    let user = fetch_user(item).await;
    let result = format !("Processed {}", user.get("name").cloned().unwrap_or_default());
    results.push(result);
   
}
results
}
#[doc = " Depyler: verified panic-free"] #[tokio::main] pub async fn main () {
    let greeting = hello_async().await;
    println !("{}", greeting);
    let service = AsyncService::new("API-Service".to_string());
    let response = service.handle_request(123).await;
    println !("{}", response);
    let items: Vec<i32>= vec ! [1, 2, 3, 4, 5];
    let batch_results = process_batch(& items).await;
    for result in batch_results.iter().cloned() {
    println !("{}", result);
   
}
} #[cfg(test)] mod tests {
    use super::*;
    use quickcheck::{
    quickcheck, TestResult };
    #[test] fn test_process_batch_examples() {
    assert_eq !(process_batch(vec ! []), vec ! []);
    assert_eq !(process_batch(vec ! [1]), vec ! [1]);
   
}
}